### PR TITLE
add Role, CustomUserDetailsService

### PR DESCRIPTION
**Pull request by Alkey**

### CODE REVIEW

**Review by jah701 — APPROVED** [review]
Well done! 🥇

File: src/main/java/com/dev/cinema/dao/impl/RoleDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+@RequiredArgsConstructor
+public class RoleDaoImpl implements RoleDao {
+    private static final Logger logger = Logger.getLogger(UserDaoImpl.class);
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/dao/impl/RoleDaoImpl.java]:
```suggestion
    private static final Logger logger = Logger.getLogger(RoleDaoImpl.class);
```

File: src/main/java/com/dev/cinema/dao/impl/RoleDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.save(role);
+            transaction.commit();
+            logger.info("Role was added to DB " + role.toString());
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't add role", e);
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/dao/impl/RoleDaoImpl.java]:
```suggestion
            throw new DataProcessingException("Can't add role " + role, e);
```

File: src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java [hunk 3]
```diff
@@ -49,7 +45,9 @@ public User add(User user) {
     public Optional<User> findByEmail(String email) {
         try (Session session = sessionFactory.openSession()) {
             Query<User> userQuery =
-                    session.createQuery("from User where email = :email", User.class);
+                    session.createQuery("from User us "
+                            + "left join fetch us.roles "
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/dao/impl/UserDaoImpl.java]: should we use left join here?

File: src/main/java/com/dev/cinema/model/User.java [hunk 4]
```diff
@@ -18,4 +20,6 @@
     @Column(unique = true)
     private String email;
     private String password;
+    @ManyToMany()
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/model/User.java]:
```suggestion
    @ManyToMany
```

File: src/main/java/com/dev/cinema/security/CustomUserDetailsService.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            builder.password(user.getPassword());
+            builder.roles(user.getRoles()
+                    .stream()
+                    .map(Role::getRoleName)
+                    .map(RoleName::name)
+                    .toArray(String[]::new));
+        } else {
+            throw new UsernameNotFoundException("Can't found user");
+        }
+        return builder.build();
```
Comment (Sofasmile) [inline src/main/java/com/dev/cinema/security/CustomUserDetailsService.java]:
```suggestion
        User userOptional = service.findByEmail(email).orElseThrow(() -> new UsernameNotFoundException("Can't found user by email " + email));
        UserBuilder builder = org.springframework.security.core.userdetails.User.withUsername(email);
        builder.password(user.getPassword());
        builder.roles(user.getRoles()
                .stream()
                .map(role -> role.getRoleName().name())
                .toArray(String[]::new));
        return builder.build();
```

**Review by Sofasmile — APPROVED** [review]

**Pull request merged**